prometheus/push.go

```diff
@@ -19,47 +19,153 @@
 
 package prometheus
 
-// Push triggers a metric collection by the default registry and pushes all
-// collected metrics to the Pushgateway specified by url. See the Pushgateway
-// documentation for detailed implications of the job and instance
-// parameter. instance can be left empty. You can use just host:port or ip:port
-// as url, in which case 'http://' is added automatically. You can also include
-// the schema in the URL. However, do not include the '/metrics/jobs/...' part.
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/prometheus/common/expfmt"
+	"github.com/prometheus/common/model"
+)
+
+// Push works like PushFrom but collects the pushed metrics from the
+// DefaultRegistry.
+func Push(job string, grouping map[string]string, url string) error {
+	return PushFrom(DefaultRegistry, job, grouping, url)
+}
+
+// PushAdd works like PushAddFrom but collects the pushed metrics from the
+// DefaultRegistry.
+func PushAdd(job string, grouping map[string]string, url string) error {
+	return PushAddFrom(DefaultRegistry, job, grouping, url)
+}
+
+// PushFrom triggers a metric collection by the provided registry and pushes all
+// collected metrics to the Pushgateway specified by url, using the provided job
+// name and the (optional) further grouping labels (the grouping map may be
+// nil). See the Pushgateway documentation for detailed implications of the job
+// and other grouping labels. Neither the job name nor any grouping label value
+// may contain a "/". The metrics pushed must not contain a job label of their
+// own nor any of the grouping labels.
+//
+// You can use just host:port or ip:port as url, in which case 'http://' is
+//added automatically. You can also include the schema in the URL. However, do
+//not include the '/metrics/jobs/...' part.
 //
-// Note that all previously pushed metrics with the same job and instance will
-// be replaced with the metrics pushed by this call. (It uses HTTP method 'PUT'
-// to push to the Pushgateway.)
-func Push(job, instance, url string) error {
-	return defRegistry.Push(job, instance, url, "PUT")
+// Note that all previously pushed metrics with the same job and other grouping
+// labels will be replaced with the metrics pushed by this call. (It uses HTTP
+// method 'PUT' to push to the Pushgateway.)
+func PushFrom(r Registry, job string, grouping map[string]string, url string) error {
```

Comment:
On 2 August 2016 at 15:04, Brian Brazil <notifications@github.com> wrote:
>
> I believe we should have these operations on particular on objects as methods, rather than as top-level functions.

Yes. I got that. Really!

I still disagree, though.

> > With the current approach, that's one call. With a struct wrapping, that becomes a two step process: Create an instance of the struct, then call a method.
>
> With custom registries it becomes two calls anyway, as you need to create the new registry whereas previously it used the default.

Most people will use the `PushCollectors` form. That would be two
instead of one calls.

With a custom registry, it would be three instead of two calls. The
instantiation of a "class" (which doesn't even exist in Go) would be
an additional step in any case.

> We could do that. What would the plan be for handling auth?

Handling auth is out of scope of this PR. (I'm not saying it's not
important. And breaking the usage only once is great, but I'm afraid
this will drag on forever if we load more changes into it.) In any
case, moving push related stuff into an own packages is at least
containing future complexity in that package.

-- 
Björn Rabenstein, Engineer
http://soundcloud.com/brabenstein

SoundCloud Ltd. | Rheinsberger Str. 76/77, 10115 Berlin, Germany
Managing Director: Alexander Ljung | Incorporated in England & Wales
with Company No. 6343600 | Local Branch Office | AG Charlottenburg  |
HRB 110657B